underfs/oss/src/main/java/alluxio/underfs/oss/OSSActiveSyncProvider.java

```diff
@@ -0,0 +1,356 @@
+
+/**
+ * OSS Active Sync Provider.
+ */
+public class OSSActiveSyncProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(OSSActiveSyncProvider.class);
+  private static final Logger SAMPLING_LOG = new SamplingLogger(LOG, 10L * Constants.MINUTE_MS);
+
+  // These read write locks protect the state (maps) managed by this class.
+  private final Lock mReadLock;
+  private final Lock mWriteLock;
+  private final ScheduledExecutorService mExecutorService;
+  private final int mActiveUfsSyncMaxActivity;
+  private final int mActiveUfsSyncMaxAge;
+
+  private final long mActiveUfsSyncInterval;
+  private Future<?> mPollingThread;
+  private List<AlluxioURI> mUfsUriList;
+
+  // a map mapping SyncPoints to a set of files that have been changed under that syncPoint
+  private Map<String, Set<AlluxioURI>> mChangedFiles;
+  // Use an integer to indicate the activity level of the sync point
+  // TODO(yuzhu): Merge the three maps into one map
+  private Map<String, Integer> mActivity;
+  private Map<String, Integer> mAge;
+  private Map<String, Long> mTxIdMap;
+  private long mCurrentTxId;
+  private boolean mEventMissed;
+
+  private final OSS mClient;
+
+  private ConcurrentMap<AlluxioURI, Long> mUfsUriLastModified;
+
+  private final String mOssFolderSuffix = "_$folder$";
+
+  /**
+   * Constructor for supported OSS Active Sync Provider.
+   *
+   * @param ufsConf   Alluxio UFS configuration
+   * @param ossClient The OSS Client
+   */
+  public OSSActiveSyncProvider(UnderFileSystemConfiguration ufsConf, OSS ossClient) {
+    mChangedFiles = new ConcurrentHashMap<>();
+    mActivity = new ConcurrentHashMap<>();
+    mAge = new ConcurrentHashMap<>();
+    ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+    mReadLock = lock.readLock();
+    mWriteLock = lock.writeLock();
+    mExecutorService = Executors.newScheduledThreadPool(
+        ufsConf.getInt(PropertyKey.MASTER_UFS_ACTIVE_SYNC_THREAD_POOL_SIZE),
+        ThreadFactoryUtils.build("OSSActiveSyncProvider-%d", false));
+    mPollingThread = null;
+    mUfsUriList = new CopyOnWriteArrayList<>();
+    mEventMissed = false;
+    mTxIdMap = new ConcurrentHashMap<>();
+    mCurrentTxId = SyncInfo.INVALID_TXID;
+    mActiveUfsSyncMaxActivity = ufsConf.getInt(PropertyKey.MASTER_UFS_ACTIVE_SYNC_MAX_ACTIVITIES);
+    mActiveUfsSyncMaxAge = ufsConf.getInt(PropertyKey.MASTER_UFS_ACTIVE_SYNC_MAX_AGE);
+    mActiveUfsSyncInterval = ufsConf.getMs(PropertyKey.MASTER_UFS_ACTIVE_SYNC_INTERVAL);
+    mClient = ossClient;
+    mUfsUriLastModified = new ConcurrentHashMap<>();
+  }
+
+  /**
+   * Start the accounting for the next window of events.
+   * <p>
+   * This includes increasing the age of the unsynced syncpoints and decrease the activity of
+   * unsynced syncpoints.
+   */
+  private void initNextWindow() {
+    for (Map.Entry<String, Integer> activity : mActivity.entrySet()) {
+      String syncPoint = activity.getKey();
+      mActivity.put(syncPoint, activity.getValue() / 10);
```

Comment:
sry, I don't understand why `activity.getValue() / 10`

Reply:
This is to avoid synchronizing a large number of files when they are changing frequently